### ISSUE TITLE
pip install fails on Ubuntu

### ISSUE THREAD

**Issue by sushanttripathy** [jakebian/quiver#22]
Pip install on Ubuntu 16.04 with Python 2.7 fails the following error
```shell
Installing collected packages: quiver-engine
  Running setup.py install for quiver-engine ... error
    Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-Yr7lN5/quiver-engine/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-XVYrCV-record/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-2.7
    creating build/lib.linux-x86_64-2.7/quiver_engine
    copying quiver_engine/util.py -> build/lib.linux-x86_64-2.7/quiver_engine
    copying quiver_engine/server.py -> build/lib.linux-x86_64-2.7/quiver_engine
    copying quiver_engine/__init__.py -> build/lib.linux-x86_64-2.7/quiver_engine
    copying quiver_engine/layer_result_generators.py -> build/lib.linux-x86_64-2.7/quiver_engine
    copying quiver_engine/imagenet_utils.py -> build/lib.linux-x86_64-2.7/quiver_engine
    running egg_info
    writing requirements to quiver_engine.egg-info/requires.txt
    writing quiver_engine.egg-info/PKG-INFO
    writing top-level names to quiver_engine.egg-info/top_level.txt
    writing dependency_links to quiver_engine.egg-info/dependency_links.txt
    warning: manifest_maker: standard file '-c' not found
    
    reading manifest file 'quiver_engine.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'quiver_engine.egg-info/SOURCES.txt'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-Yr7lN5/quiver-engine/setup.py", line 20, in <module>
        'pillow'
      File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/dist-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run
        self.run_command('build')
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
        self.run_command(cmd_name)
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 52, in run
        self.build_package_data()
      File "/usr/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 107, in build_package_data
        for package, src_dir, build_dir, filenames in self.data_files:
      File "/usr/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 65, in __getattr__
        self.data_files = self._get_data_files()
      File "/usr/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 79, in _get_data_files
        return list(map(self._get_pkg_data_files, self.packages or ()))
      File "/usr/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 91, in _get_pkg_data_files
        for file in self.find_data_files(package, src_dir)
      File "/usr/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 98, in find_data_files
        + self.package_data.get(package, []))
    TypeError: can only concatenate list (not "str") to list
    
    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-Yr7lN5/quiver-engine/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-XVYrCV-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-Yr7lN5/quiver-engine/
```

**Comment by ramakrishna-git** [comment]
same error on ubuntu 14.04 and python 2.7

writing manifest file 'quiver_engine.egg-info/SOURCES.txt'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-tSKij6-build/setup.py", line 21, in <module>   
        'pillow'
      File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run
        self.run_command('build')
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
        self.run_command(cmd_name)
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 52, in run
        self.build_package_data()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 107, in build_package_data
        for package, src_dir, build_dir, filenames in self.data_files:
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 65, in __getattr__
        self.data_files = self._get_data_files()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 79, in _get_data_files
        return list(map(self._get_pkg_data_files, self.packages or ()))
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 91, in _get_pkg_data_files
        for file in self.find_data_files(package, src_dir)
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 98, in find_data_files
        + self.package_data.get(package, []))
    TypeError: can only concatenate list (not "str") to list

    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-tSKij6-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-bW_Wnk-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-tSKij6-build/

**Comment by ramakrishna-git** [comment]
any workaround for this issue?

**Comment by sushanttripathy** [comment]
I updated to latest keras. That seemed to fix the issue. Hope this helps.

**Comment by ramakrishna-git** [comment]
I updated to Keras 1.2.0 but still get the same error.

Here is the what i get when I try to install quiver_engine on Ubuntu 14.04.5 LTS (docker linux)

pip install quiver_engine
Collecting quiver_engine
  Downloading quiver_engine-0.1.4.1.4.tar.gz (398kB)   
    100% |################################| 399kB 655kB/s
Requirement already satisfied (use --upgrade to upgrade): keras in /usr/local/lib/python2.7/dist-packages (from quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): tensorflow in /usr/local/lib/python2.7/dist-packages (from quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): flask in /usr/local/lib/python2.7/dist-packages (from quiver_engine)
Collecting flask_cors (from quiver_engine)
  Downloading Flask_Cors-3.0.2-py2.py3-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): gevent in /usr/local/lib/python2.7/dist-packages (from quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): numpy in /usr/local/lib/python2.7/dist-packages (from quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): pillow in /usr/local/lib/python2.7/dist-packages (from quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): theano in /src/theano (from keras->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): pyyaml in /usr/local/lib/python2.7/dist-packages/PyYAML-3.12-py2.7-linux-x86_64.egg (from keras->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): six in /usr/local/lib/python2.7/dist-packages (from keras->quiver_engine)
Collecting protobuf==3.0.0 (from tensorflow->quiver_engine)
  Downloading protobuf-3.0.0-py2.py3-none-any.whl (342kB)
    100% |################################| 348kB 990kB/s
Requirement already satisfied (use --upgrade to upgrade): wheel in /usr/local/lib/python2.7/dist-packages (from tensorflow->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): mock>=2.0.0 in /usr/local/lib/python2.7/dist-packages (from tensorflow->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): itsdangerous>=0.21 in /usr/local/lib/python2.7/dist-packages (from flask->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): Werkzeug>=0.7 in /usr/local/lib/python2.7/dist-packages (from flask->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): Jinja2>=2.4 in /usr/local/lib/python2.7/dist-packages (from flask->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): greenlet>=0.4.9 in /usr/local/lib/python2.7/dist-packages (from gevent->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): scipy>=0.14 in /usr/local/lib/python2.7/dist-packages (from theano->keras->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): setuptools in /usr/local/lib/python2.7/dist-packages (from protobuf==3.0.0->tensorflow->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): funcsigs>=1; python_version < "3.3" in /usr/local/lib/python2.7/dist-packages (from mock>=2.0.0->tensorflow->quiver_e
ngine)
Requirement already satisfied (use --upgrade to upgrade): pbr>=0.11 in /usr/local/lib/python2.7/dist-packages (from mock>=2.0.0->tensorflow->quiver_engine)
Requirement already satisfied (use --upgrade to upgrade): MarkupSafe in /usr/local/lib/python2.7/dist-packages (from Jinja2>=2.4->flask->quiver_engine)
Building wheels for collected packages: quiver-engine  
  Running setup.py bdist_wheel for quiver-engine ... error
  Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-BZ75Np/quiver-engine/setup.py';exec(compile(getattr(tokenize, 'open'
, open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" bdist_wheel -d /tmp/tmpxT7yhupip-wheel- --python-tag cp27:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-2.7
  creating build/lib.linux-x86_64-2.7/quiver_engine
  copying quiver_engine/util.py -> build/lib.linux-x86_64-2.7/quiver_engine
  copying quiver_engine/imagenet_utils.py -> build/lib.linux-x86_64-2.7/quiver_engine
  copying quiver_engine/layer_result_generators.py -> build/lib.linux-x86_64-2.7/quiver_engine
  copying quiver_engine/server.py -> build/lib.linux-x86_64-2.7/quiver_engine
copying quiver_engine/__init__.py -> build/lib.linux-x86_64-2.7/quiver_engine
    running egg_info
    writing requirements to quiver_engine.egg-info/requires.txt
    writing quiver_engine.egg-info/PKG-INFO
    writing top-level names to quiver_engine.egg-info/top_level.txt
    writing dependency_links to quiver_engine.egg-info/dependency_links.txt
    warning: manifest_maker: standard file '-c' not found

    reading manifest file 'quiver_engine.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'quiver_engine.egg-info/SOURCES.txt'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-BZ75Np/quiver-engine/setup.py", line 21, in <module>
        'pillow'
      File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python2.7/distutils/command/install.py", line 601, in run
        self.run_command('build')
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/lib/python2.7/distutils/command/build.py", line 128, in run
        self.run_command(cmd_name)
      File "/usr/lib/python2.7/distutils/cmd.py", line 326, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 52, in run
        self.build_package_data()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 107, in build_package_data
        for package, src_dir, build_dir, filenames in self.data_files:
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 65, in __getattr__
        self.data_files = self._get_data_files()
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 79, in _get_data_files
        return list(map(self._get_pkg_data_files, self.packages or ()))

File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 91, in _get_pkg_data_files
        for file in self.find_data_files(package, src_dir)
      File "/usr/local/lib/python2.7/dist-packages/setuptools/command/build_py.py", line 98, in find_data_files
        + self.package_data.get(package, []))
    TypeError: can only concatenate list (not "str") to list

    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-BZ75Np/quiver-engine/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-2vM6ES-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-BZ75Np/quiver-engine/
You are using pip version 8.1.2, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

**Comment by ramakrishna-git** [comment]
Update:
It worked after following steps mentioned in similar issue thread for windows.
Steps:
1. clone the git repository locally.
2. Replace 'quiverboard/dist/*' with ['quiverboard/dist/*'] in setup.py
3. Installed with
      python setup.py install

The above problem got resolved.
I faced another issue after the above steps, server was throwing 500 Internal server error while requesting /inputs url failing at code trying to jsonify list of files.
Solution: Upgraded flask to 0.12 which adds the required jsonify(array) support.

Last issue faced was related to error while inferring input shape from model. Hardcoded the value single_input_shape = [200, 200] to make it work.

Cheers,
